### ISSUE TITLE
Possible miscomputed ABC size

### ISSUE THREAD

**Issue by pdxwolfy** [bbatsov/rubocop#3395]
I think there may be a computation error in computing the ABC size for the following code:

```ruby
def update_score_buffer!(cards_in_play, score_buffer, who)
  if who == "dealer" || who == "both"
    score_buffer[:dealer_int_1_card] = total_int(cards_in_play[:dealer][0])
    score_buffer[:dealer_str_1_card] = total_str(cards_in_play[:dealer][0])
    score_buffer[:dealer_int] = total_int(cards_in_play[:dealer])
    score_buffer[:dealer_str] = total_str(cards_in_play[:dealer])
  end
  if who == "player" || who == "both"
    score_buffer[:player_int] = total_int(cards_in_play[:player])
    score_buffer[:player_str] = total_str(cards_in_play[:player])
  end
end
```

```
xxx.rb:104:1: C: Metrics/AbcSize: Assignment Branch Condition size for build_value_row is too high. [18.14/18]  (http://c2.com/cgi/wiki?AbcMetric)
```

Not my code - it's a student's, and I was trying to use it to show him how to deal with Abc size complaints. By my computations, there are 6 assignments, 20 branches, and 4 conditions in the above, which should give an ABC size of 21.26. But, my version of rubocop says 18.14, which I can only make work by assuming that there are only 18 branches. (I count 4 branches on each of lines 3 and 4, and 3 each on lines 5, 6, 9, and 10.)

--------

## Expected behavior

Issue following complaint:

```
xxx.rb:104:1: C: Metrics/AbcSize: Assignment Branch Condition size for build_value_row is too high. [21.26/18]  (http://c2.com/cgi/wiki?AbcMetric)
```

## Actual behavior

```
xxx.rb:104:1: C: Metrics/AbcSize: Assignment Branch Condition size for build_value_row is too high. [18.14/15]  (http://c2.com/cgi/wiki?AbcMetric)
```

## Steps to reproduce the problem

```
$ rubocop -c /usr/local/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubocop-0.42.0/config/enabled.yml xxx.rb
```

## RuboCop version

```
$ rubocop -V
0.42.0 (using Parser 2.3.1.2, running on ruby 2.3.1 x86_64-darwin15)
```

**Comment by jonas054** [comment]
The abstract syntax tree of the method is as follows:
```
s(:def, :update_score_buffer!,
  s(:args,
    s(:arg, :cards_in_play),
    s(:arg, :score_buffer),
    s(:arg, :who)),
  s(:begin,
    s(:if,
      s(:or,
        s(:send,
          s(:lvar, :who), :==,
          s(:str, "dealer")),
        s(:send,
          s(:lvar, :who), :==,
          s(:str, "both"))),
      s(:begin,
        s(:send,
          s(:lvar, :score_buffer), :[]=,
          s(:sym, :dealer_int_1_card),
          s(:send, nil, :total_int,
            s(:send,
              s(:send,
                s(:lvar, :cards_in_play), :[],
                s(:sym, :dealer)), :[],
              s(:int, 0)))),
        s(:send,
          s(:lvar, :score_buffer), :[]=,
          s(:sym, :dealer_str_1_card),
          s(:send, nil, :total_str,
            s(:send,
              s(:send,
                s(:lvar, :cards_in_play), :[],
                s(:sym, :dealer)), :[],
              s(:int, 0)))),
        s(:send,
          s(:lvar, :score_buffer), :[]=,
          s(:sym, :dealer_int),
          s(:send, nil, :total_int,
            s(:send,
              s(:lvar, :cards_in_play), :[],
              s(:sym, :dealer)))),
        s(:send,
          s(:lvar, :score_buffer), :[]=,
          s(:sym, :dealer_str),
          s(:send, nil, :total_str,
            s(:send,
              s(:lvar, :cards_in_play), :[],
              s(:sym, :dealer))))), nil),
    s(:if,
      s(:or,
        s(:send,
          s(:lvar, :who), :==,
          s(:str, "player")),
        s(:send,
          s(:lvar, :who), :==,
          s(:str, "both"))),
      s(:begin,
        s(:send,
          s(:lvar, :score_buffer), :[]=,
          s(:sym, :player_int),
          s(:send, nil, :total_int,
            s(:send,
              s(:lvar, :cards_in_play), :[],
              s(:sym, :player)))),
        s(:send,
          s(:lvar, :score_buffer), :[]=,
          s(:sym, :player_str),
          s(:send, nil, :total_str,
            s(:send,
              s(:lvar, :cards_in_play), :[],
              s(:sym, :player))))), nil)))
```
Adding some printing to the cop, I produced this:
```
condition: if who == "dealer" || who == "both"
    score_buffer[:dealer_int_1_card] = total_int(cards_in_play[:dealer][0])
    score_buffer[:dealer_str_1_card] = total_str(cards_in_play[:dealer][0])
    score_buffer[:dealer_int] = total_int(cards_in_play[:dealer])
    score_buffer[:dealer_str] = total_str(cards_in_play[:dealer])
  end
condition: who == "dealer" || who == "both"
branch: who == "dealer"
branch: who == "both"
assignment: score_buffer[:dealer_int_1_card] = total_int(cards_in_play[:dealer][0])
branch: total_int(cards_in_play[:dealer][0])
branch: cards_in_play[:dealer][0]
branch: cards_in_play[:dealer]
assignment: score_buffer[:dealer_str_1_card] = total_str(cards_in_play[:dealer][0])
branch: total_str(cards_in_play[:dealer][0])
branch: cards_in_play[:dealer][0]
branch: cards_in_play[:dealer]
assignment: score_buffer[:dealer_int] = total_int(cards_in_play[:dealer])
branch: total_int(cards_in_play[:dealer])
branch: cards_in_play[:dealer]
assignment: score_buffer[:dealer_str] = total_str(cards_in_play[:dealer])
branch: total_str(cards_in_play[:dealer])
branch: cards_in_play[:dealer]
condition: if who == "player" || who == "both"
    score_buffer[:player_int] = total_int(cards_in_play[:player])
    score_buffer[:player_str] = total_str(cards_in_play[:player])
  end
condition: who == "player" || who == "both"
branch: who == "player"
branch: who == "both"
assignment: score_buffer[:player_int] = total_int(cards_in_play[:player])
branch: total_int(cards_in_play[:player])
branch: cards_in_play[:player]
assignment: score_buffer[:player_str] = total_str(cards_in_play[:player])
branch: total_str(cards_in_play[:player])
branch: cards_in_play[:player]
```
And the total counts for A, B, and C are:
```
{:assignment=>6, :branch=>18, :condition=>4}
```
Does this help you identify where your expectation and the behavior of `Metrics/AbcSize` deviate?

**Comment by pdxwolfy** [comment]
yes it does, thank you.
